core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala

```diff
@@ -0,0 +1,552 @@
+
+  override def equals(other: Any): Boolean = other match {
+    case that: DriverSubmission =>
+      (that canEqual this) &&
+        submissionId == that.submissionId
+    case _ => false
+  }
+}
+
+private [spark] case class ClusterTaskState(
+    val submission: DriverSubmission,
+    val taskId: TaskID,
+    val slaveId: SlaveID,
+    var taskState: Option[TaskStatus],
+    var driverState: DriverState,
+    var startDate: Date,
+    val lastRetry: Option[RetryState] = None) extends Serializable {
+
+  def copy(): ClusterTaskState = {
+    ClusterTaskState(submission, taskId, slaveId, taskState, driverState, startDate, lastRetry)
+  }
+}
+
+private[spark] case class SubmitResponse(id: String, success: Boolean, message: String)
+
+private[spark] case class StatusResponse(
+    id: String,
+    success: Boolean,
+    state: String,
+    status: Option[TaskStatus] = None)
+
+private[spark] case class KillResponse(id: String, success: Boolean, message: String)
+
+private[spark] case class ClusterSchedulerState(
+    appId: String,
+    queuedDrivers: Iterable[DriverSubmission],
+    launchedDrivers: Iterable[ClusterTaskState],
+    finishedDrivers: Iterable[ClusterTaskState],
+    retryList: Iterable[RetryState])
+
+private[spark] trait ClusterScheduler {
+  def submitDriver(desc: MesosDriverDescription): SubmitResponse
+
+  def killDriver(submissionId: String): KillResponse
+
+  def getStatus(submissionId: String): StatusResponse
+
+  def getState(): ClusterSchedulerState
+}
+
+private[spark] class MesosClusterScheduler(
+    engineFactory: ClusterPersistenceEngineFactory,
+    conf: SparkConf) extends Scheduler with MesosSchedulerHelper with ClusterScheduler {
+
+  var frameworkUrl: String = _
+
+  val master = conf.get("spark.master")
+  val appName = conf.get("spark.app.name")
+  val queuedCapacity = conf.getInt("spark.deploy.mesos.queuedDrivers", 200)
+  val retainedDrivers = conf.getInt("spark.deploy.retainedDrivers", 200)
+  val maxRetryWaitTime = conf.getInt("spark.mesos.cluster.retry.wait.max", 60) // 1 minute
+  val state = engineFactory.createEngine("scheduler")
+  val stateTimeout =
+    Duration.create(conf.getLong("spark.mesos.cluster.recover.timeout", 30), "seconds")
+
+  val stateLock = new ReentrantLock()
+
+  val finishedDrivers = new mutable.ArrayBuffer[ClusterTaskState](retainedDrivers)
+
+  val nextDriverNumber: AtomicLong = new AtomicLong(0)
+  var appId: String = null
+
+  private var launchedDrivers: LaunchedDrivers = _
+
+  private var queue: DriverQueue = _
+
+  def createDateFormat = new SimpleDateFormat("yyyyMMddHHmmss")  // For application IDs
+
+  private var superviseRetryList: SuperviseRetryList = _
+
+  private def newDriverId(submitDate: Date): String = {
+    "driver-%s-%04d".format(
+        createDateFormat.format(submitDate), nextDriverNumber.incrementAndGet())
+  }
+
+  def submitDriver(desc: MesosDriverDescription): SubmitResponse = {
```

Comment:
It's highly unlikely that we will have another submission API for mesos cluster mode other than REST. I think we can just return `SubmitRestProtocolResponse` to avoid creating all these new case classes here.